source/builtin_macros/src/syntax.rs

```diff
@@ -3981,6 +4002,70 @@ pub(crate) fn sig_specs_attr(
     visitor.take_sig_specs(&mut spec, ret_pat, sig.constness.is_some(), sig_span)
 }
 
+pub(crate) fn while_loop_spec_attr(
+    erase_ghost: EraseGhost,
+    spec_attr: syn_verus::LoopSpec,
+) -> Vec<Stmt> {
+    let mut visitor = Visitor {
+        erase_ghost,
+        use_spec_traits: true,
+        inside_ghost: 1,
+        inside_type: 0,
+        inside_external_code: 0,
+        inside_const: false,
+        inside_arith: InsideArith::None,
+        assign_to: false,
+        rustdoc: env_rustdoc(),
+    };
+    let mut spec_attr = spec_attr;
+    visitor.visit_loop_spec(&mut spec_attr);
+    let syn_verus::LoopSpec {
+        invariants,
+        invariant_except_breaks,
+        invariant_ensures,
+        ensures,
+        decreases,
+        ..
+    } = spec_attr;
+    let mut stmt = vec![];
+    visitor.add_loop_specs(
+        &mut stmt,
+        invariant_except_breaks,
+        invariants,
+        invariant_ensures,
+        ensures,
+        decreases,
+    );
+    stmt
+}
+
+pub(crate) fn for_loop_spec_attr(
+    erase_ghost: EraseGhost,
+    spec_attr: syn_verus::LoopSpec,
+    forloop: syn::ExprForLoop,
+) -> syn_verus::Expr {
+    let mut visitor = Visitor {
+        erase_ghost,
+        use_spec_traits: true,
+        inside_ghost: 1,
+        inside_type: 0,
+        inside_external_code: 0,
+        inside_const: false,
+        inside_arith: InsideArith::None,
+        assign_to: false,
+        rustdoc: env_rustdoc(),
+    };
+    let mut spec_attr = spec_attr;
+    visitor.visit_loop_spec(&mut spec_attr);
+    let syn_verus::LoopSpec { invariants, decreases, .. } = spec_attr;
+    let mut forloop: ExprForLoop =
+        syn_verus::parse2(forloop.to_token_stream().into()).expect("ExprForLoop");
```

Comment:
This parses the `for` loop's executable code with syn_verus, which we're trying to avoid.  Instead of calling syn_verus::parse2, I think we can construct a syn_verus::ExprForLoop directly by supplying the fields for pat, expr, body, etc. directly.  These fields can be constructed without calling the parser by using syn_verus::Expr::Verbatim and syn_verus::Pat::Verbatim.